### ISSUE TITLE
RTCQuicStream state

### ISSUE THREAD

**Issue by shampson** [w3c/webrtc-quic#65]
Is it necessary to have the RTCQuicStream states exposed? What a web developer needs to know:
 - When the stream becomes available for reading/writing. This doesn't need a state if issues [54](https://github.com/w3c/webrtc-quic/issues/54) and [50](https://github.com/w3c/webrtc-quic/issues/50) get resolved. If you get a stream it immediately starts out readable and writable (or just one if unidirectional).
 - When the stream is no longer readable. This is solved currently with a synchronous call to readInto returning a negative number, or possibly adding a FIN bool to the waitForReadable promise (issue [63](https://github.com/w3c/webrtc-quic/issues/63))
- When the stream is no longer writable. The developer knows this because either:
  - They called finish()/reset()
  - Currently onstatechange event is fired because the stream's state transitions to closed due to the remote side calling reset(). This is really the only asynchronous event we need for the stream.

I would propose just having [[Readable]] and [[Writable]] exposed (issue [64](https://github.com/w3c/webrtc-quic/issues/64)) and an event when the remote side indicates that the local side should no longer write (onremotereset event?). This would simplify things, and the developer wouldn't need to understand that onstatechange for the stream only occurs for a remote reset() call (issue [56](https://github.com/w3c/webrtc-quic/issues/56)).

Thoughts? I'm curious if there are reasons I'm missing for exposing the stream states.

**Comment by pthatcherg** [comment]
If you just see readable/writable, then how do you know if it's because it isn't yet readable/writable (because, say, the QUIC handshake isn't complete) or because something got finished/reset?  There's kind of a before state, a steady state, and an after state.  readable/writable only helps you know if it's steady or not.

I suppose we could make readable and writable enums of something like "not yet", "yes", and "not any more".  


We should probably ask ourselves what we want the waitForX to be when you create a stream.  waitForReadable(amount) still works because it will just block until something is readable, regardless of state.  But we may have messed up waitForWritable in this sense because waitForWriteBufferAmountBelow() will resolve immediately even if you're not writable/open yet.  So you'd have to wait for the state change and then call waitForWriteBufferAmountBelow.  It would be more convenient to just call waitForWritable(amount) without having to worry about the state change.  But then we're back to re-adding setMaxWriteBufferedAmount, which we just removed.  But maybe that's OK. Maybe it was a mistake to remove it.

In other words, can we make it so users don't have to check any of these state transitions and just use waitForReadable and waitForWritable?

**Comment by shampson** [comment]
If we resolve [issue 50](https://github.com/w3c/webrtc-quic/issues/50) to require that the Transport must be connected before calling createStream(), then streams will immediately be readable AND writable. I clarified why in [issue 54](https://github.com/w3c/webrtc-quic/issues/54). 

Just using waitForReadable and waitForWritable is an interesting idea (no events for state changes at all). Once things are no longer readable/writable then any call to waitForReadable/waitForWritable would just be rejected.
